Groundwork for mypy support in FuzzManager

nth10sd opened the pull request:
This PR lays the groundwork for mypy support in FuzzManager.

- In the `misc/` folder, I had to rename `afl-libfuzzer` to `afl_libfuzzer` as mypy does not like dashes
- I took the mypy config in [Lithium](https://github.com/MozillaSecurity/lithium/) and adapted it here
- I added mypy and other `types-*` packages to the `requirements3.0.txt` file

GitHub Actions CI [is here](https://github.com/nth10sd/FuzzManager/actions).

Current master branch has 3188 mypy errors after installing all the *-stubs packages (and their latest versions on a Python 3.10 venv), this PR now has 1558 errors left. (51.13% out of 3188 fixed)

nth10sd commented:
@jschwartzentruber This may be a starting point. Running `mypy` on the root directory shows:

`Found 2950 errors in 161 files (checked 208 source files)`

Using the following command, we may be able to (on CI) check whether more mypy regressions are creeping in:

```
python -c 'import subprocess; out = subprocess.run(["mypy"], stdout=subprocess.PIPE); known_num = int(out.stdout.splitlines()[-1].decode("utf-8").split("Found ", maxsplit=1)[1].split(" errors in ", maxsplit=1)[0]); print(f"new mypy error total: {known_num}" if known_num > 2950 else "");'
```

This prints a blank line if the number is not exceeded, otherwise it prints `new mypy error total: {known_num}`

(update the 2950 figure with the count from mypy as preferred)

nth10sd commented:
Rev a0e50ae and rev 4d8582d bring the count down to 2711.

Feel free to start whittling them down!

nth10sd commented:
Currently down to 2689, and rebased to tip.

nth10sd commented:
More recent progress and after rebasing to tip: `new mypy error total: 2453`

nth10sd commented:
Note to self: These may be good starting points.

```
    136  error: Call to untyped function "ProgramConfiguration" in typed context  [no-untyped-call]
     76  error: Call to untyped function "assert_contains" in typed context  [no-untyped-call]
     51  error: Call to untyped function "create_config" in typed context  [no-untyped-call]
     49  error: Call to untyped function "matches" in typed context  [no-untyped-call]
     42  error: Call to untyped function "create_pool" in typed context  [no-untyped-call]
     39  error: Call to untyped function "CrashSignature" in typed context  [no-untyped-call]
     23  error: Call to untyped function "get_or_create_restricted" of "User" in typed context  [no-untyped-call]
     21  error: Call to untyped function "_crashentry_create" in typed context  [no-untyped-call]
     18  error: Call to untyped function "create_instance" in typed context  [no-untyped-call]
     16  error: Call to untyped function "__init__" in typed context  [no-untyped-call]
     13  error: Call to untyped function "_update_pool_status" in typed context  [no-untyped-call]
     13  error: Call to untyped function "_check_regex_matches" in typed context  [no-untyped-call]
     11  error: Call to untyped function "get_instance" of "CloudProvider" in typed context  [no-untyped-call]
     11  error: Call to untyped function "_create_user" in typed context  [no-untyped-call]
     10  error: Call to untyped function "main" in typed context  [no-untyped-call]
     10  error: Call to untyped function "checkRevisionsEquivalent" in typed context  [no-untyped-call]
     10  error: Call to untyped function "_connect" in typed context  [no-untyped-call]
     10  error: Call to untyped function "StringMatch" in typed context  [no-untyped-call]
      9  error: Call to untyped function "_update_provider_status" in typed context  [no-untyped-call]
      8  error: Call to untyped function "get_query_obj" in typed context  [no-untyped-call]
      8  error: Call to untyped function "check_authorized_for_crash_entry" in typed context  [no-untyped-call]
      8  error: Call to untyped function "SimplePersistentApplication" in typed context  [no-untyped-call]
      7  error: Call to untyped function "post" in typed context  [no-untyped-call]
      7  error: Call to untyped function "create_task" in typed context  [no-untyped-call]
      7  error: Call to untyped function "Collector" in typed context  [no-untyped-call]
      7  error: Call to untyped function "CloudProviderTemporaryFailure" in typed context  [no-untyped-call]
      6  error: Call to untyped function "getArrayChecked" in typed context  [no-untyped-call]
      6  error: Call to untyped function "check_authorized_for_signature" in typed context  [no-untyped-call]
      6  error: Call to untyped function "_check" in typed context  [no-untyped-call]
      5  error: Call to untyped function "select_better" in typed context  [no-untyped-call]
      5  error: Call to untyped function "get_flattened_names" in typed context  [no-untyped-call]
      5  error: Call to untyped function "get" in typed context  [no-untyped-call]
      5  error: Call to untyped function "fromBinary" of "ProgramConfiguration" in typed context  [no-untyped-call]
      5  error: Call to untyped function "filter_crash_entries_by_toolfilter" in typed context  [no-untyped-call]
      5  error: Call to untyped function "apply_include_exclude_directives" in typed context  [no-untyped-call]
      5  error: Call to untyped function "_cache_parent" in typed context  [no-untyped-call]
      5  error: Call to untyped function "__getTypeChecked" in typed context  [no-untyped-call]
      5  error: Call to untyped function "EC2Reporter" in typed context  [no-untyped-call]
      4  error: Call to untyped function "testRevision" in typed context  [no-untyped-call]
      4  error: Call to untyped function "print_indent" in typed context  [no-untyped-call]
      4  error: Call to untyped function "getParents" in typed context  [no-untyped-call]
      4  error: Call to untyped function "getNumberOrStringChecked" in typed context  [no-untyped-call]
      4  error: Call to untyped function "deny_restricted_users" in typed context  [no-untyped-call]
      4  error: Call to untyped function "deferRawFields" of "CrashEntry" in typed context  [no-untyped-call]
      4  error: Call to untyped function "calculate_summary_fields" in typed context  [no-untyped-call]
      4  error: Call to untyped function "_render_node" in typed context  [no-untyped-call]
      4  error: Call to untyped function "_match" in typed context  [no-untyped-call]
      4  error: Call to untyped function "_compare_rest_result_to_bucket" in typed context  [no-untyped-call]
      4  error: Call to untyped function "NumberMatch" in typed context  [no-untyped-call]
      3  error: Call to untyped function "write_aggregated_stats_libfuzzer" in typed context  [no-untyped-call]
      3  error: Call to untyped function "setup_firefox" in typed context  [no-untyped-call]
      3  error: Call to untyped function "sanitize_utf8" in typed context  [no-untyped-call]
      3  error: Call to untyped function "requests_retry" in typed context  [no-untyped-call]
      3  error: Call to untyped function "matchRequiresTest" in typed context  [no-untyped-call]
      3  error: Call to untyped function "get_serializer" in typed context  [no-untyped-call]
      3  error: Call to untyped function "getStringChecked" in typed context  [no-untyped-call]
      3  error: Call to untyped function "getSectionMap" in typed context  [no-untyped-call]
      3  error: Call to untyped function "getObjectOrStringChecked" in typed context  [no-untyped-call]
      3  error: Call to untyped function "getCrashInfo" in typed context  [no-untyped-call]
      3  error: Call to untyped function "deferRawFields" in typed context  [no-untyped-call]
      3  error: Call to untyped function "convert_num" in typed context  [no-untyped-call]
      3  error: Call to untyped function "_write_log_test" in typed context  [no-untyped-call]
      3  error: Call to untyped function "_is_dir" in typed context  [no-untyped-call]
      3  error: Call to untyped function "_diff" in typed context  [no-untyped-call]
      3  error: Call to untyped function "_crashed" in typed context  [no-untyped-call]
      3  error: Call to untyped function "_compare_rest_result_to_crash" in typed context  [no-untyped-call]
      3  error: Call to untyped function "_compare_created_data_to_crash" in typed context  [no-untyped-call]
      3  error: Call to untyped function "_LowercaseDict" in typed context  [no-untyped-call]
      3  error: Call to untyped function "TaskStatusReporter" in typed context  [no-untyped-call]
      3  error: Call to untyped function "HGSourceCodeProvider" in typed context  [no-untyped-call]
      3  error: Call to untyped function "CrashEntryVueSerializer" in typed context  [no-untyped-call]
      2  error: Call to untyped function "write_stats_file" in typed context  [no-untyped-call]
      2  error: Call to untyped function "version_info_from_coverage_data" of "CovReporter" in typed context  [no-untyped-call]
      2  error: Call to untyped function "strip" of "Collection" in typed context  [no-untyped-call]
      2  error: Call to untyped function "start" in typed context  [no-untyped-call]
      2  error: Call to untyped function "save" in typed context  [no-untyped-call]
      2  error: Call to untyped function "response" in typed context  [no-untyped-call]
      2  error: Call to untyped function "release" in typed context  [no-untyped-call]
      2  error: Call to untyped function "recurse_get_dependent_configurations" in typed context  [no-untyped-call]
      2  error: Call to untyped function "processInput" in typed context  [no-untyped-call]
      2  error: Call to untyped function "preprocess_coverage_data" of "CovReporter" in typed context  [no-untyped-call]
      2  error: Call to untyped function "merge_recursive" in typed context  [no-untyped-call]
      2  error: Call to untyped function "json_to_query" in typed context  [no-untyped-call]
      2  error: Call to untyped function "handle_repository" in typed context  [no-untyped-call]
      2  error: Call to untyped function "get_or_create_pool" in typed context  [no-untyped-call]
      2  error: Call to untyped function "get_options" in typed context  [no-untyped-call]
      2  error: Call to untyped function "get_labels" in typed context  [no-untyped-call]
      2  error: Call to untyped function "get_datasets" in typed context  [no-untyped-call]
      2  error: Call to untyped function "get_data_colors" in typed context  [no-untyped-call]
      2  error: Call to untyped function "get_colors" in typed context  [no-untyped-call]
      2  error: Call to untyped function "getSource" in typed context  [no-untyped-call]
      2  error: Call to untyped function "getSignature" in typed context  [no-untyped-call]
      2  error: Call to untyped function "getDiffLocations" of "Utils" in typed context  [no-untyped-call]
      2  error: Call to untyped function "getBugs" in typed context  [no-untyped-call]
      2  error: Call to untyped function "fromBinaryArgs" of "AutoRunner" in typed context  [no-untyped-call]
      2  error: Call to untyped function "filter_signatures_by_toolfilter" in typed context  [no-untyped-call]
      2  error: Call to untyped function "filter_bucket_hits_by_toolfilter" in typed context  [no-untyped-call]
      2  error: Call to untyped function "diff" in typed context  [no-untyped-call]
      2  error: Call to untyped function "decrement_count" of "BucketHit" in typed context  [no-untyped-call]
      2  error: Call to untyped function "decrement_count" in typed context  [no-untyped-call]
      2  error: Call to untyped function "create_combined_coverage" of "CovReporter" in typed context  [no-untyped-call]
      2  error: Call to untyped function "command_file_to_list" in typed context  [no-untyped-call]
      2  error: Call to untyped function "apply_transform" in typed context  [no-untyped-call]
      2  error: Call to untyped function "annotate_delta" in typed context  [no-untyped-call]
      2  error: Call to untyped function "add_children" in typed context  [no-untyped-call]
      2  error: Call to untyped function "addResponsePrefix" in typed context  [no-untyped-call]
      2  error: Call to untyped function "_wait_child_stopped" in typed context  [no-untyped-call]
      2  error: Call to untyped function "_update_task_run" in typed context  [no-untyped-call]
      2  error: Call to untyped function "_node_to_instance" in typed context  [no-untyped-call]
      2  error: Call to untyped function "_match" of "StackFramesSymptom" in typed context  [no-untyped-call]
      2  error: Call to untyped function "_compare_rest_result_to_template" in typed context  [no-untyped-call]
      2  error: Call to untyped function "__validate" in typed context  [no-untyped-call]
      2  error: Call to untyped function "__upload_queue_files" in typed context  [no-untyped-call]
      2  error: Call to untyped function "__handleConfigPOST" in typed context  [no-untyped-call]
      2  error: Call to untyped function "__get_machine_id" in typed context  [no-untyped-call]
      2  error: Call to untyped function "__get_flattened_names" in typed context  [no-untyped-call]
      2  error: Call to untyped function "__apply_include_exclude_directives" in typed context  [no-untyped-call]
      2  error: Call to untyped function "TestCase" in typed context  [no-untyped-call]
      2  error: Call to untyped function "StreamCollector" in typed context  [no-untyped-call]
      2  error: Call to untyped function "StackFramesSymptom" in typed context  [no-untyped-call]
      2  error: Call to untyped function "RedisLock" in typed context  [no-untyped-call]
      2  error: Call to untyped function "PoolConfiguration" in typed context  [no-untyped-call]
      2  error: Call to untyped function "LibFuzzerMonitor" in typed context  [no-untyped-call]
      2  error: Call to untyped function "ConfigurationFiles" in typed context  [no-untyped-call]
      2  error: Call to untyped function "Collection" in typed context  [no-untyped-call]
      2  error: Call to untyped function "BugzillaREST" in typed context  [no-untyped-call]
      1  error: Call to untyped function "write_aggregated_stats_afl" in typed context  [no-untyped-call]
      1  error: Call to untyped function "version_info_from_coverage_data" in typed context  [no-untyped-call]
      1  error: Call to untyped function "test_binary_asan" in typed context  [no-untyped-call]
      1  error: Call to untyped function "strip" in typed context  [no-untyped-call]
      1  error: Call to untyped function "scan_crashes" in typed context  [no-untyped-call]
      1  error: Call to untyped function "renderError" in typed context  [no-untyped-call]
      1  error: Call to untyped function "remove_childrens_children" of "Collection" in typed context  [no-untyped-call]
      1  error: Call to untyped function "preprocess_coverage_data" in typed context  [no-untyped-call]
      1  error: Call to untyped function "paginated" in typed context  [no-untyped-call]
      1  error: Call to untyped function "migration_hook_result" in typed context  [no-untyped-call]
      1  error: Call to untyped function "merge_coverage_data" in typed context  [no-untyped-call]
      1  error: Call to untyped function "login" in typed context  [no-untyped-call]
      1  error: Call to untyped function "isCyclic" in typed context  [no-untyped-call]
      1  error: Call to untyped function "increment_count" of "BucketHit" in typed context  [no-untyped-call]
      1  error: Call to untyped function "increment_count" in typed context  [no-untyped-call]
      1  error: Call to untyped function "hg" of "_result" in typed context  [no-untyped-call]
      1  error: Call to untyped function "has_permission" in typed context  [no-untyped-call]
      1  error: Call to untyped function "git" of "_result" in typed context  [no-untyped-call]
      1  error: Call to untyped function "get_spot_prices" in typed context  [no-untyped-call]
      1  error: Call to untyped function "get_spot_price_per_region" in typed context  [no-untyped-call]
      1  error: Call to untyped function "get_queryset" in typed context  [no-untyped-call]
      1  error: Call to untyped function "get_price_median" in typed context  [no-untyped-call]
      1  error: Call to untyped function "get_price" in typed context  [no-untyped-call]
      1  error: Call to untyped function "get_instance_types" in typed context  [no-untyped-call]
      1  error: Call to untyped function "getUnifiedDiff" in typed context  [no-untyped-call]
      1  error: Call to untyped function "getTestcase" in typed context  [no-untyped-call]
      1  error: Call to untyped function "getSymptomsDiff" in typed context  [no-untyped-call]
      1  error: Call to untyped function "getInstance" in typed context  [no-untyped-call]
      1  error: Call to untyped function "getBugStatus" in typed context  [no-untyped-call]
      1  error: Call to untyped function "getBug" in typed context  [no-untyped-call]
      1  error: Call to untyped function "getASanTrace" in typed context  [no-untyped-call]
      1  error: Call to untyped function "fromJSONObject" of "Symptom" in typed context  [no-untyped-call]
      1  error: Call to untyped function "flatten" in typed context  [no-untyped-call]
      1  error: Call to untyped function "fit" in typed context  [no-untyped-call]
      1  error: Call to untyped function "create_weekly_report_mc" in typed context  [no-untyped-call]
      1  error: Call to untyped function "create_repository" of "_result" in typed context  [no-untyped-call]
      1  error: Call to untyped function "create_poolmsg" in typed context  [no-untyped-call]
      1  error: Call to untyped function "create_collection_file" of "_result" in typed context  [no-untyped-call]
      1  error: Call to untyped function "create_bugprovider" of "_cm_result" in typed context  [no-untyped-call]
      1  error: Call to untyped function "connect_to_region" in typed context  [no-untyped-call]
      1  error: Call to untyped function "acquire" in typed context  [no-untyped-call]
      1  error: Call to untyped function "_start_pool_instances" in typed context  [no-untyped-call]
      1  error: Call to untyped function "_get_skus_paginated" in typed context  [no-untyped-call]
      1  error: Call to untyped function "_diff" of "StackFramesSymptom" in typed context  [no-untyped-call]
      1  error: Call to untyped function "_determine_best_location" in typed context  [no-untyped-call]
      1  error: Call to untyped function "_create_laniakea_images" in typed context  [no-untyped-call]
      1  error: Call to untyped function "_compare_rest_result_to_bugprovider" in typed context  [no-untyped-call]
      1  error: Call to untyped function "_MockReq" in typed context  [no-untyped-call]
      1  error: Call to untyped function "_MockInstance" in typed context  [no-untyped-call]
      1  error: Call to untyped function "TestcaseSymptom" in typed context  [no-untyped-call]
      1  error: Call to untyped function "TaskClusterConsumer" in typed context  [no-untyped-call]
      1  error: Call to untyped function "StackSizeSymptom" in typed context  [no-untyped-call]
      1  error: Call to untyped function "StackFrameSymptom" in typed context  [no-untyped-call]
      1  error: Call to untyped function "RecurseReportSummaryTree" in typed context  [no-untyped-call]
      1  error: Call to untyped function "RecurseConfigTree" in typed context  [no-untyped-call]
      1  error: Call to untyped function "OutputSymptom" in typed context  [no-untyped-call]
      1  error: Call to untyped function "InstructionSymptom" in typed context  [no-untyped-call]
      1  error: Call to untyped function "GITSourceCodeProvider" in typed context  [no-untyped-call]
      1  error: Call to untyped function "GDBRunner" in typed context  [no-untyped-call]
      1  error: Call to untyped function "CrashEntrySerializer" in typed context  [no-untyped-call]
      1  error: Call to untyped function "CrashAddressSymptom" in typed context  [no-untyped-call]
      1  error: Call to untyped function "CovReporter" in typed context  [no-untyped-call]
      1  error: Call to untyped function "ConfigurationFile" in typed context  [no-untyped-call]
      1  error: Call to untyped function "CloudProviderInstanceCountError" in typed context  [no-untyped-call]
      1  error: Call to untyped function "CloudProviderError" in typed context  [no-untyped-call]
      1  error: Call to untyped function "ASanRunner" in typed context  [no-untyped-call]
```

nth10sd commented:
Current master branch has 3183 mypy errors after installing all the *-stubs packages, this PR now has 2360 errors left. (25.86% out of 3183 fixed)

sylvestre commented:
Impressive work! bravo

nth10sd commented:
Current master branch has 3188 mypy errors after installing all the *-stubs packages (and their latest versions on a Python 3.10 venv), this PR now has 1740 errors left. (45.42% out of 3188 fixed)

nth10sd commented:
For now, CI is running on [a new repo belonging to me](https://github.com/nth10sd/typed-fuzzmanager/actions).

nth10sd commented:
Superseded by #829.

nth10sd commented:
@jschwartzentruber some questions:

1. Do you still want to import `annotations` now that we are on Python 3.7+? Or do you prefer old-style type annotations? (e.g. `Optional[str]`)
1. For dictionaries, are you ok with `dict[str, Any]` typing?
1. For *args and **kwargs, are you ok with `Any` types?
1. For the items that I am not sure, do you prefer me to put `Any` or you'd like to fill in the rest?
1. I still need to fix things up back to where I was, just prior to merging your changes in

This will take some time, and I have a work trip ahead that may take out the next few weeks for me.

jschwartzentruber commented:
> 1. Are you ok to import `annotations` now that we are on Python 3.7+? Or do you prefer old-style type annotations? (e.g. `Optional[str]`)

yes, since we dropped 3.6 I think using the future import is fine. 

> 2. For dictionaries, are you ok with `dict[str, Any]` typing?

yes

> 3. For *args and **kwargs, are you ok with `Any` types?

yes

> 4. For the items that I am not sure (e.g. decorator stuff), do you prefer me to put `Any` or you'd like to fill in the rest?

`Any` is fine

nth10sd commented:
```
$ python -u -m tox -e mypy -- .
mypy develop-inst-noop: /home/skyxubu/FuzzManager
mypy installed: adal==1.2.7,aiohttp==3.8.1,aiosignal==1.2.0,amqp==2.6.1,apache-libcloud==2.4.0,appdirs==1.4.4,async-timeout==4.0.2,asynctest==0.13.0,attrs==21.4.0,azure-common==1.1.14,azure-mgmt-nspkg==3.0.2,azure-mgmt-resource==1.2.2,azure-nspkg==3.0.2,billiard==3.6.4.0,boto==2.49.0,boto3==1.21.3,botocore==1.24.3,celery==4.4.7,certifi==2021.10.8,cffi==1.15.0,charset-normalizer==2.0.12,coreapi==2.3.3,coreschema==0.0.4,coverage==6.3.1,cryptography==36.0.1,Deprecated==1.2.13,Django==2.2.27,django-chartjs==1.3,django-crispy-forms==1.11.2,django-enumfields==2.1.1,django-model-utils==4.2.0,django-notifications-hq==1.6.0,django-stubs==1.10.1,django-stubs-ext==0.4.0,djangorestframework==3.11.2,djangorestframework-stubs==1.5.0,fasteners==0.17.3,frozenlist==1.3.0,fuzzing-decision @ git+https://github.com/MozillaSecurity/orion@079c2eca4555998982f7f1459d252b2e047efc94#subdirectory=services/fuzzing-decision,-e git+ssh://git@github.com/nth10sd/FuzzManager.git@6593c254acd0686b444b0c66bb5ab478db98ae3b#egg=FuzzManager,idna==3.3,importlib-metadata==4.11.1,iniconfig==1.1.1,isodate==0.6.1,itypes==1.2.0,jeepney==0.7.1,Jinja2==3.1.2,jmespath==0.10.0,jsonfield==3.1.0,keyring==23.5.0,kombu==4.6.11,laniakea==1.17.1,MarkupSafe==2.1.1,mohawk==1.1.0,MozillaPulse==1.3,msrest==0.6.21,msrestazure==0.4.34,multidict==6.0.2,mypy==0.931,mypy-extensions==0.4.3,oauthlib==3.2.0,packaging==21.3,packet-python==1.37.1,pluggy==1.0.0,py==1.11.0,pycparser==2.21,pycryptodome==3.7.3,PyJWT==2.3.0,pyparsing==3.0.7,pytest==6.2.5,pytest-cov==3.0.0,pytest-django==4.5.2,pytest-mock==3.7.0,pytest-pythonpath==0.7.4,python-dateutil==2.8.2,pytz==2021.3,PyYAML==6.0,redis==4.1.4,requests==2.27.1,requests-oauthlib==1.3.1,s3transfer==0.5.1,SecretStorage==3.3.1,six==1.16.0,slugid==2.0.0,sqlparse==0.4.2,swapper==1.3.0,taskcluster==44.13.0,taskcluster-urls==13.0.1,toml==0.10.2,tomli==2.0.1,typed-ast==1.5.3,types-Markdown==3.3.14,types-pytz==2021.3.7,types-PyYAML==6.0.7,types-requests==2.27.25,types-urllib3==1.26.14,typing_extensions==4.1.1,uritemplate==4.1.1,urllib3==1.26.8,vine==1.3.0,whitenoise==5.2.0,wrapt==1.13.3,yarl==1.7.2,zipp==3.7.0
mypy run-test-pre: PYTHONHASHSEED='3039526982'
mypy run-test: commands[0] | mypy --install-types --non-interactive .
Error constructing plugin instance of NewSemanalDjangoPlugin

Traceback (most recent call last):
  File "/home/skyxubu/FuzzManager/.tox/mypy/bin/mypy", line 8, in <module>
    sys.exit(console_entry())
  File "/home/skyxubu/FuzzManager/.tox/mypy/lib/python3.7/site-packages/mypy/__main__.py", line 12, in console_entry
    main(None, sys.stdout, sys.stderr)
  File "mypy/main.py", line 96, in main
  File "mypy/main.py", line 173, in run_build
  File "mypy/build.py", line 180, in build
  File "mypy/build.py", line 231, in _build
  File "mypy/build.py", line 478, in load_plugins
  File "mypy/build.py", line 456, in load_plugins_from_config
  File "/home/skyxubu/FuzzManager/.tox/mypy/lib/python3.7/site-packages/mypy_django_plugin/main.py", line 70, in __init__
    self.django_context = DjangoContext(self.plugin_config.django_settings_module)
  File "/home/skyxubu/FuzzManager/.tox/mypy/lib/python3.7/site-packages/mypy_django_plugin/django/context.py", line 96, in __init__
    apps, settings = initialize_django(self.django_settings_module)
  File "/home/skyxubu/FuzzManager/.tox/mypy/lib/python3.7/site-packages/mypy_django_plugin/django/context.py", line 80, in initialize_django
    apps.populate(settings.INSTALLED_APPS)
  File "/home/skyxubu/FuzzManager/.tox/mypy/lib/python3.7/site-packages/django/apps/registry.py", line 91, in populate
    app_config = AppConfig.create(entry)
  File "/home/skyxubu/FuzzManager/.tox/mypy/lib/python3.7/site-packages/django/apps/config.py", line 90, in create
    module = import_module(entry)
  File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/skyxubu/FuzzManager/server/ec2spotmanager/__init__.py", line 3, in <module>
    from . import tasks  # noqa
  File "/home/skyxubu/FuzzManager/server/ec2spotmanager/tasks.py", line 15, in <module>
    from . import cron  # noqa ensure cron tasks get registered
  File "/home/skyxubu/FuzzManager/server/ec2spotmanager/cron.py", line 16, in <module>
    from .CloudProvider.CloudProvider import INSTANCE_STATE, PROVIDERS, CloudProvider
  File "/home/skyxubu/FuzzManager/server/ec2spotmanager/CloudProvider/CloudProvider.py", line 25, in <module>
    from ec2spotmanager.models import PoolConfiguration
  File "/home/skyxubu/FuzzManager/server/ec2spotmanager/models.py", line 44, in <module>
    class PoolConfiguration(models.Model):
  File "/home/skyxubu/FuzzManager/.tox/mypy/lib/python3.7/site-packages/django/db/models/base.py", line 103, in __new__
    app_config = apps.get_containing_app_config(module)
  File "/home/skyxubu/FuzzManager/.tox/mypy/lib/python3.7/site-packages/django/apps/registry.py", line 252, in get_containing_app_config
    self.check_apps_ready()
  File "/home/skyxubu/FuzzManager/.tox/mypy/lib/python3.7/site-packages/django/apps/registry.py", line 135, in check_apps_ready
    raise AppRegistryNotReady("Apps aren't loaded yet.")
django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
ERROR: InvocationError for command /home/skyxubu/FuzzManager/.tox/mypy/bin/mypy --install-types --non-interactive . (exited with code 1)
_________________________________________________________________________________ summary __________________________________________________________________________________
ERROR:   mypy: commands failed
```

@jschwartzentruber As requested by you, I switched to tox, and removed the casting, but I hit this error I cannot seem to fix. Do you mind unblocking this? (And also bumping mypy to the latest version while you're at it)

nth10sd commented:
Note that the tox setup edits `server/server/settings.py` before running `mypy`, then reverts them to make mypy and its plugins happy. Suggestions welcome on how this should be resolved.

Before proceeding any further, @jschwartzentruber can you replicate this setup with `tox -e mypy -- .`? It'll be great if I can hand this off.
